lib/parsers/acorn-extensions.js

```diff
@@ -0,0 +1,191 @@
+      // export * as ns from '...'
+      specifier.exported = this.parseIdent(true);
+      node.specifiers = [
+        this.finishNode(specifier, "ExportNamespaceSpecifier")
+      ];
+      this.parseExportSpecifiersMaybe(node);
+      this.parseExportFromWithCheck(node, exports, true);
+    } else {
+      // export * from '...'
+      this.parseExportFromWithCheck(node, exports, true);
+      return this.finishNode(node, "ExportAllDeclaration");
+    }
+  } else if (this.isExportDefaultSpecifier()) {
+    // export def from '...'
+    const specifier = this.startNode();
+    specifier.exported = this.parseIdent(true);
+    node.specifiers = [
+      this.finishNode(specifier, "ExportDefaultSpecifier")
+    ];
+    if (this.type === tt.comma &&
+        this.lookAhead(1).type === tt.star) {
+      // export def, * as ns from '...'
+      this.expect(tt.comma);
+      const specifier = this.startNode();
+      this.expect(tt.star);
+      this.expectContextual("as");
+      specifier.exported = this.parseIdent(true);
+      node.specifiers.push(
+        this.finishNode(specifier, "ExportNamespaceSpecifier")
+      );
+    } else {
+      // export def, { x, y as z } from '...'
+      this.parseExportSpecifiersMaybe(node);
+    }
+    this.parseExportFromWithCheck(node, exports, true);
+  } else if (this.eat(tt._default)) {
+    // export default ...
+    this.checkExport(exports, "default", this.lastTokStart);
+    let isAsync;
+    if (this.type === tt._function || (isAsync = this.isAsyncFunction())) {
+      let fNode = this.startNode();
+      this.next();
+      if (isAsync) this.next();
+      node.declaration = this.parseFunction(fNode, "nullableID", false, isAsync);
+    } else if (this.type === tt._class) {
+      let cNode = this.startNode();
+      node.declaration = this.parseClass(cNode, "nullableID");
+    } else {
+      node.declaration = this.parseMaybeAssign();
+      this.semicolon();
+    }
+    return this.finishNode(node, "ExportDefaultDeclaration");
+  } else if (this.shouldParseExportDeclaration()) {
+    // export var|const|let|function|class ...
+    node.declaration = this.parseStatement(true);
+    if (node.declaration.type === "VariableDeclaration") {
+      this.checkVariableExport(exports, node.declaration.declarations);
+    } else {
+      this.checkExport(
+        exports,
+        node.declaration.id.name,
+        node.declaration.id.start
+      );
+    }
+    node.specifiers = [];
+    node.source = null;
+  } else {
+    // export { x, y as z } [from '...']
+    node.declaration = null;
+    node.specifiers = this.parseExportSpecifiers(exports);
+    this.parseExportFrom(node, false);
+  }
+  return this.finishNode(node, "ExportNamedDeclaration");
+}
+
+function lookAhead(n) {
+  const old = Object.assign(Object.create(null), this);
+  while (n-- > 0) this.nextToken();
+  const copy = Object.assign(Object.create(null), this);
+  Object.assign(this, old);
+  return copy;
```

Comment:
We're taking a snapshot of the parser so that we can roll it back after advancing its state to look at later tokens.